framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceScenes.java

```diff
@@ -452,6 +460,44 @@ public void updateInferredMethodReturnType(
         AnnotatedTypeMirror rhsATM = atf.getAnnotatedType(retNode.getTree().getExpression());
         helper.updateAnnotationSetInScene(
                 method.returnType, atf, jaifPath, rhsATM, lhsATM, TypeUseLocation.RETURN);
+
+        // Now, update overridden methods based on the implementation we just saw.
+        // If the overridden method is abstract, then WPI cannot update it directly
+        // (because no calls directly resolve to it). This inference is similar to
+        // the inference procedure for method parameters: both are updated based
+        // only on the implementations (in this case) or call-sites (for method
+        // parameters) that are available to WPI.
+        for (Map.Entry<AnnotatedDeclaredType, ExecutableElement> pair :
+                overriddenMethods.entrySet()) {
+
+            AnnotatedDeclaredType superclassType = pair.getKey();
+            ExecutableElement superclassMethodDef = pair.getValue();
+
+            AnnotatedExecutableType overriddenMethod =
+                    AnnotatedTypes.asMemberOf(
+                            atf.getProcessingEnv().getTypeUtils(),
+                            atf,
+                            superclassType,
+                            superclassMethodDef);
+
+            if (overriddenMethod.getElement().getModifiers().contains(Modifier.ABSTRACT)) {
```

Comment:
Why aren't facts about return values propagated to all overridden methods?

Reply:
Because when I wrote this I was only interested in abstract methods. I've fixed it.